util/src/main/java/io/kubernetes/client/Exec.java

```diff
@@ -191,6 +199,135 @@ public Process exec(
   }
 
+  /**
+   * A convenience method. Executes a command remotely on a pod and monitors for events in that
+   * execution. The monitored events are: <br>
+   * - connection established (onOpen) <br>
+   * - connection closed (onClosed) <br>
+   * - execution error occurred (onError) <br>
+   * This method also allows to specify a MAX timeout for the execution and returns a future in
+   * order to monitor the execution flow.
+   *
+   * @param namespace a namespace the target pod "lives" in
+   * @param podName a name of the pod to exec the command on
+   * @param onOpen a callback invoked upon the connection established event.
+   * @param onClosed a callback invoked upon the process termination. Return code might not always
+   *     be there.
+   * @param onError a callback to handle k8s errors (NOT the command errors/stderr!)
+   * @param timeoutMs timeout in milliseconds for the execution. I.e. the execution will take this
+   *     many ms or less. If the timeout command is running longer than the allowed timeout, the
+   *     command will be "asked" to terminate gracefully. If the command is still running after the
+   *     grace period, the sigkill will be issued. If null is passed, the timeout will not be used
+   *     and will wait for process to exit itself.
+   * @param tty whether you need tty to pipe the data. TTY mode will trim some binary data in order
+   *     to make it possible to show on screen (tty)
+   * @param command a tokenized command to run on the pod
+   * @return a {@link Future} promise representing this execution. Unless something goes south, the
+   *     promise will contain the process return exit code. If the timeoutMs is non-null and the
+   *     timeout expires before the process exits, promise will contain {@link Integer#MAX_VALUE}.
+   * @throws IOException
+   */
+  public Future<Integer> exec(
+      String namespace,
+      String podName,
+      Consumer<IOTrio> onOpen,
+      BiConsumer<Integer, IOTrio> onClosed,
+      BiConsumer<Throwable, IOTrio> onError,
+      Long timeoutMs,
+      boolean tty,
+      String... command)
+      throws IOException {
+    CompletableFuture<Integer> future = new CompletableFuture<>();
+    IOTrio io = new IOTrio();
+    String cmdStr = Arrays.toString(command);
+    try {
+      Process process =
+          newExecutionBuilder(namespace, podName, command)
+              .setStdin(true)
+              .setStdout(true)
+              .setStderr(true)
+              .setTty(tty)
+              .execute();
+      io.onClose(
+          (code, timeout) -> {
+            process.destroy();
+            waitForProcessToExit(process, timeout, cmdStr, err -> onError.accept(err, io));
+            // processWaitingThread will handle the rest
+          });
+      io.setStdin(process.getOutputStream());
+      io.setStdout(process.getInputStream());
+      io.setStderr(process.getErrorStream());
+      Thread processWaitingThread =
+          new Thread(
```

Comment:
This method is going to be hard to test because of this. Can you use an `ExecutorService` instead? That can be injected in the test.

Reply:
Sure thing!

```
  private void runAsync(String taskName, Runnable task) {
    if (this.executorService == null) {
      Thread thread = new Thread(task);
      thread.setName(taskName);
      thread.start();
    } else {
      executorService.submit(task);
    }
  }
```

Will this cut it? Using an `ExecutorService` if there is supplied one, and falling back to standalone Thread if there isn't one. And a simple Exec.setExecutorService() to supply the executor. Although the executor would also be used in this sole method (unless reused somewhere else later on).

I really don't feel like adding yet another parameter to the method, since there already are 8+ and the ES would be mostly optional.

Another option - use a protected field. or even a `protected runAsync(String taskName, Runnable task){...}`, which could be overridden in tests. Which approach would you prefer?

Reply:
This approach is fine. Thanks!

Reply:
which is "this"? :) 
I have gone with the `protected runAsync(String taskName, Runnable task){...}` (seemed more appropriate). Shall I leave it or replace it with the code block in my comment above?